### PR TITLE
[FEATURE REQUEST] Update space image

### PR DESCRIPTION
**Pull request by joragua**
## Related Issues
App: https://github.com/owncloud/android/issues/4691

- [x] Add changelog files for the fixed issues in folder changelog/unreleased. More info [here](https://github.com/owncloud/android/tree/master/changelog#create-changelog-items)
- [x] Add feature to Release Notes in `ReleaseNotesViewModel.kt` creating a new `ReleaseNote()` with String resources (if required)

_____

## QA

Test plan:

https://github.com/owncloud/QA/blob/master/Mobile/Android/Executions/Release_4.7/Spaces%20-%20Edit%20image.md

Reports:

- [x] (1) Lack of error message when no permissions https://github.com/owncloud/android/pull/4705#issuecomment-3490196619 [FIXED]
- [X] (2) Space image not updated after recovering connection https://github.com/owncloud/android/pull/4705#issuecomment-3490672287  [[WONT FIX]](https://github.com/owncloud/android/pull/4705#issuecomment-3492313716)

### CODE REVIEW

File: owncloudApp/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -851,6 +851,9 @@
     <string name="edit_space">Edit space</string>
     <string name="edit_space_correctly">Space updated correctly</string>
     <string name="edit_space_failed">Space could not be updated</string>
+    <string name="edit_space_image">Edit space image</string>
+    <string name="edit_space_image_correctly">Space image updated correctly</string>
+    <string name="edit_space_image_failed">Space image could not be updated</string>
```
Comment (jesmrec) [inline owncloudApp/src/main/res/values/strings.xml]: I'd go for a shorter one: `Edit image` . Web and iOS clients go for that approach. Then, if we go for it, the snackbar messages should be shortened as well

`Image updated correctly`
`Image could not be updated`

File: owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/SpacesListFragment.kt [hunk 2]
```diff
@@ -210,6 +237,17 @@ class SpacesListFragment :
             showSpaceMenuOptionsDialog(menuOptions)
         }
 
+        collectLatestLifecycleFlow(transfersViewModel.transfersWithSpaceStateFlow) { transfersWithSpace->
```
Comment (jesmrec) [inline owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/SpacesListFragment.kt]: missing blank between `transfersWithSpace` and `->` ?

File: owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/SpacesListFragment.kt [hunk 3]
```diff
@@ -210,6 +237,17 @@ class SpacesListFragment :
...
+            val remotePath = SPACE_CONFIG_DIR + selectedImageName
+            val matchedTransfer = transfersWithSpace.map { it.first }.find { it.remotePath == remotePath }
+
+            if (matchedTransfer != null && matchedTransfer.status == TransferStatus.TRANSFER_SUCCEEDED &&
+                lastUpdatedRemotePath != matchedTransfer.remotePath && lastUpdatedRemotePath != null) {
```
Comment (jesmrec) [inline owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/SpacesListFragment.kt]: 💯  to avoid repetitions

File: owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/SpacesListFragment.kt [hunk 4]
```diff
@@ -382,6 +420,14 @@ class SpacesListFragment :
                             negativeButtonText = getString(R.string.common_no)
                         )
                     }
+                    SpaceMenuOption.EDIT_IMAGE -> {
+                        val action = Intent(Intent.ACTION_OPEN_DOCUMENT).apply {
+                            addCategory(Intent.CATEGORY_OPENABLE)
+                            type = "image/*"
+                            putExtra(Intent.EXTRA_MIME_TYPES, arrayOf("image/jpeg", "image/png", "image/bmp", "image/x-ms-bmp", "image/gif"))
```
Comment (jesmrec) [inline owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/SpacesListFragment.kt]: Should this static values be moved to a const or companion object?

File: owncloudData/src/test/java/com/owncloud/android/data/spaces/repository/OCSpacesRepositoryTest.kt [hunk 5]
```diff
@@ -350,6 +351,31 @@ class OCSpacesRepositoryTest {
...
+
+        verify(exactly = 1) {
+            remoteSpacesDataSource.editSpaceImage(
+                accountName = OC_ACCOUNT_NAME,
+                spaceId = OC_SPACE_PROJECT_WITH_IMAGE.id,
+                imageId = OC_SPACE_SPECIAL_IMAGE.id
+            )
+        }
+    }
+
```
Comment (jesmrec) [inline owncloudData/src/test/java/com/owncloud/android/data/spaces/repository/OCSpacesRepositoryTest.kt:378]: Tests added are OK, just as a question: would it make sense to add negative tests? i mean, to check the correct response if the action failed (image editing returned an error, a null or so)

File: owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/SpacesListFragment.kt [hunk 6]
```diff
@@ -382,6 +420,14 @@ class SpacesListFragment :
                             negativeButtonText = getString(R.string.common_no)
                         )
                     }
+                    SpaceMenuOption.EDIT_IMAGE -> {
+                        val action = Intent(Intent.ACTION_OPEN_DOCUMENT).apply {
```
Comment (jesmrec) [inline owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/SpacesListFragment.kt:437]: As a question: selecting just one file is the default behaviour?

**Review by jesmrec — CHANGES_REQUESTED** [review]

File: owncloudApp/src/main/res/values/strings.xml (hunk 1 above)
Comment (joragua) [reply owncloudApp/src/main/res/values/strings.xml]: Ok! Let's go for it 😄

File: owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/SpacesListFragment.kt (hunk 2 above)
Comment (joragua) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/SpacesListFragment.kt]: Added 👍🏻

File: owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/SpacesListFragment.kt (hunk 6 above)
Comment (joragua) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/SpacesListFragment.kt:437]: Yes. If you want to allow multiple selections, just add this flag: `putExtra(Intent.EXTRA_ALLOW_MULTIPLE, true)`

File: owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/SpacesListFragment.kt (hunk 4 above)
Comment (joragua) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/SpacesListFragment.kt]: Yeah! Created new constants for these values ✅

File: owncloudData/src/test/java/com/owncloud/android/data/spaces/repository/OCSpacesRepositoryTest.kt (hunk 5 above)
Comment (joragua) [reply owncloudData/src/test/java/com/owncloud/android/data/spaces/repository/OCSpacesRepositoryTest.kt:378]: In this case, the two methods covered by the new tests don’t return null or throw errors. The ones that can are the network operations and use cases, but their tests haven’t been updated yet.

File: owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/SpacesListFragment.kt (hunk 6 above)
Comment (jesmrec) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/SpacesListFragment.kt:437]: perfect 👍

File: owncloudData/src/test/java/com/owncloud/android/data/spaces/repository/OCSpacesRepositoryTest.kt (hunk 5 above)
Comment (jesmrec) [reply owncloudData/src/test/java/com/owncloud/android/data/spaces/repository/OCSpacesRepositoryTest.kt:378]: 👍

File: owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/SpacesListFragment.kt (hunk 2 above)
Comment (jesmrec) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/SpacesListFragment.kt]: Fixed

File: owncloudApp/src/main/res/values/strings.xml (hunk 1 above)
Comment (jesmrec) [reply owncloudApp/src/main/res/values/strings.xml]: Done 🥇

File: owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/SpacesListFragment.kt (hunk 4 above)
Comment (jesmrec) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/spaces/SpacesListFragment.kt]: Cool! perfect!

**Review by jesmrec — APPROVED** [review]

**Comment by jesmrec** [comment]
Moving to QA 🚀

**Comment by jesmrec** [comment]
### (1) [FIXED]

1. User1 is manager of the space
2. User1 choose `Edit image`
3. Before choosing an image, another space manager downgrades User1 to `viewer`
4. User1 submit the image

**Current**:

No feedback in the app, there is no image change and no error message (just in the uploads view)

**Expected:**

As other handled errors, an snackbar would tell the user about the issue

NOTE: theoretically, it should work the same by changing to `editor`, but, backend allows the image update. IMO, this is a bug in the backend. I will create an issue there -> https://github.com/owncloud/ocis/issues/11794

Xiaomi Redmi 13, Android 15
`0749466fb`

**Comment by jesmrec** [comment]
### (2) [[WONT FIX]](https://github.com/owncloud/android/pull/4705#issuecomment-3492313716)

1. A user with permissions click on `Edit image`
2. Remove device connection
3. Pick a image -> enqueued 
4. Recover device connection

**Current:**

Upload is done, image is uploaded, but space image is not updated. There is no `PATCH` request after the successful `PUT`

**Expected:**

Space image updated.

Xiaomi Redmi 13, Android 15
`0749466fb`

**Comment by jesmrec** [comment]
Just as a comment: i detected a 7MB limit for the space image. A larger image will not update, but, a success is returned. Pointing to another server issue...

https://github.com/owncloud/ocis/issues/11832

**Comment by joragua** [comment]
**(1)** should be fixed with the last commit @jesmrec

**Comment by joragua** [comment]
About **[(2)](https://github.com/owncloud/android/pull/4705#issuecomment-3490672287)**, there are two different scenarios:

### 1st scenario

1) A user with permissions click on `Edit image`
2) Remove device connection
3) Pick a image -> `QUEUED`
4) Recover device connection
5) Stay in the spaces list

In this case, the space image will be updated automatically once the upload is completed after the connection is recovered.

### 2nd scenario
1) A user with permissions click on `Edit image`
2) Remove device connection
3) Pick a image -> `QUEUED`
4) Navigate to other view
5) Recover device connection

In this case, the image will be uploaded to the `.space` directory but the space image won't be updated. This happens because the `selectedImageName` variable (inside `SpacesListFragment`) is reset while navigating. As a result, once you return to the spaces list, the remote transfer path will not match with the selected image name and the `PATCH` operation isn't triggered. Due to technical limitations related to navigation, fragment lifecycle, and activity handling, I don’t have a clear solution for this at the moment. Anyway, it's a corner case and there is no data-loss, so probably we can assume this problem until we have a potential fix. What do you think, @jesmrec? 🤔 

#### Potential solutions discarded

1) When a transfer has the status `QUEUED` or `FAILED` the `PATCH` operation isn't performed. **Problem:** every transfer goes through the `QUEUED` status before `IN_PROGRESS` regardless of whether the connection is lost or not.

**Comment by jesmrec** [comment]
About the message just above ⬆️ ⬆️ 

In order to reproduce the problem:

1. User has initially connection, that allows to click on `Edit image`
2. Then, user loses connection while the image is chosen
3. After submitting, user must change the view, otherwise the problem is not reproducible.

Problem: image not updated

User must repeat the operation under connection (low impact side-effect)

As @joragua said, there is no critical problem at all. But the solution carries a big complexity to achieve a small piece of improvement. So, i'd go to mark it as `known problem`.

**Comment by jesmrec** [comment]
Approved from my side. 

Known problem: https://github.com/owncloud/android/pull/4705#issuecomment-3492313716
